auth0/v3/management/rest.py

```diff
@@ -1,16 +1,87 @@
-import sys
-import json
 import base64
+import json
+import sys
+import time
+
 import requests
+
 from ..exceptions import Auth0Error
 
 
-class RestClient(object):
+class NopRateLimiter(object):
+    def update(self, headers):
+        pass
+
+    def ensure_limit(self):
+        pass
+
+
+class RateLimiter(object):
+    def __init__(self):
+        self.ratelimit_reset = int(time.time()) - 1
+        self.ratelimit_remaining = 1
+
+    def update(self, headers):
+        if 'x-ratelimit-remaining' in headers:
+            self.ratelimit_remaining = int(headers['x-ratelimit-remaining'])
+
+        if 'x-ratelimit-reset' in headers:
+            self.ratelimit_reset = int(headers['x-ratelimit-reset'])
+
+    def ensure_limit(self):
+        self.ratelimit_remaining -= 1
+
+        if self.ratelimit_remaining < 0:
+            sleep_time = self.ratelimit_reset - int(time.time())
+            if sleep_time > 0:
+                time.sleep(sleep_time)
+
+
+def extract_content(func):
+    def _decorator(self, *args, **kwargs):
+        response = func(self, *args, **kwargs)
+
+        text = json.loads(response.text) if response.text else {}
+
+        if isinstance(text, dict) and 'errorCode' in text:
+            raise Auth0Error(status_code=text['statusCode'],
+                             error_code=text['errorCode'],
+                             message=text['message'])
+        return text
+
+    return _decorator
+
+
+rate_limit_max_retries = 5
+
+
+def rate_limited(func):
+    def _decorator(self, *args, **kwargs):
+        self.rate_limiter.ensure_limit()
```

Comment:
Since `time.sleep()` is going to block, what's the use on having a `for 0-5` to retry it?

Reply:
If there are parallel accesses to a rate limited API the rate limit information from a previous call may be not up to date. Retrying in case of rate limit errors should be done to overcome this.